lib/rucio/client/uploadclient.py

```diff
@@ -139,15 +139,32 @@ def upload(self, items, summary_file_path=None):
             file_did = {'scope': file['did_scope'], 'name': file['did_name']}
             dataset_did_str = file.get('dataset_did_str')
 
+            rse = file['rse']
+            rse_settings = self.rses[rse]
+            is_deterministic = rse_settings.get('deterministic', True)
+            if not is_deterministic and not pfn:
+                logger.warning('PFN has to be defined for NON-DETERMINISTIC site.')
+                continue
+            if pfn:
```

Comment:
you don't need this `if pfn` here. You made sure that a PFN is given if `not is_deterministic` evaluates to True .

Reply:
there is a bug, but different one. Actually, if no_register shouldn't be here. But otherwise, I moved this part form _collect_and_validate here. We have to be sure, that if pfn was given, no_register is set to True (but only in cases of non-deterministic sites)